File: 09-Graph/06-Cycle-Detection/cycle_detection_undirected.py
n = int(input("Enter number of vertices: "))
e = int(input("Enter number of edges: "))

graph = [[] for _ in range(n)]

print("Enter edges (u v):")

for _ in range(e):
    u, v = map(int, input().split())
    graph[u].append(v)
    graph[v].append(u)

visited = [False] * n


def has_cycle(node, parent):
    visited[node] = True

    for neighbor in graph[node]:
        if not visited[neighbor]:
            if has_cycle(neighbor, node):
                return True
        elif neighbor != parent:
            return True

    return False


cycle = False

for i in range(n):
    if not visited[i]:
        if has_cycle(i, -1):
            cycle = True
            break

if cycle:
    print("Cycle detected!")
else:
    print("No cycle detected.")

File: 09-Graph/06-Cycle-Detection/cycle_detection_undirected.c
#include <stdio.h>

#define MAX 100

int graph[MAX][MAX];
int visited[MAX];
int n;

int has_cycle(int node, int parent) {
    visited[node] = 1;

    for(int i = 0; i < n; i++) {
        if(graph[node][i]) {
            if(!visited[i]) {
                if(has_cycle(i, node))
                    return 1;
            }
            else if(i != parent) {
                return 1;
            }
        }
    }

    return 0;
}

int main() {
    int e;

    printf("Enter number of vertices: ");
    scanf("%d", &n);

    printf("Enter number of edges: ");
    scanf("%d", &e);

    printf("Enter edges (u v):\n");

    for(int i = 0; i < e; i++) {
        int u, v;

        scanf("%d %d", &u, &v);

        graph[u][v] = 1;
        graph[v][u] = 1;
    }

    int cycle = 0;

    for(int i = 0; i < n; i++) {
        if(!visited[i]) {
            if(has_cycle(i, -1)) {
                cycle = 1;
                break;
            }
        }
    }

    if(cycle)
        printf("Cycle detected!\n");
    else
        printf("No cycle detected.\n");

    return 0;
}